File: 250-500/496-next-greater-element-I.cpp
vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
    unordered_map<int, int> next;
    stack<int> s;
    
    for (const auto &num : nums2) {
        while (!s.empty() && s.top() < num) {
            int pre = s.top();
            s.pop();
            next[pre] = num;
        }
        s.push(num);
    }
    vector<int> ans;
    for (const auto &num : nums1) {
        ans.push_back((next.count(num) == 0) ? -1 : next[num]);
    }
    return ans;
}

File: python version/496-next-greater-element-I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        hash_dict = {}
        for num in nums2:
            while stack and stack[-1] < num:
                prev = stack.pop(-1)
                hash_dict[prev] = num
            stack.append(num)
        
        ans = []
        for num in nums1:
            ans.append(hash_dict.get(num, -1))
        return ans
